### PR TITLE
Toggle for build output

### PR DESCRIPTION
**Pull request by rodrmoya**
This introduces a service for feature switch, so that it's easy to enable/disable experimental features, and uses it to conditionally enable structured build output
> [Fixes  VSTS #701567](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/701567)

### CODE REVIEW

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace MonoDevelop.Core
+{
+	/// <summary>
+	/// Base class for feature switch conditions.
+	/// </summary>
+	public abstract class FeatureSwitchCondition
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: Could be testable.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+			return true;
+		}
+	}
+
+	/// <summary>
+	/// Allows aggregation of several switch conditions into a single condition.
+	/// </summary>
+	public class AggregatedFeatureSwitchCondition : FeatureSwitchCondition
+	{
+		readonly List<FeatureSwitchCondition> conditions = new List<FeatureSwitchCondition> ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: Why not just an immutable array here?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+			this.allMustPass = allMustPass;
+		}
+
+		internal override bool Evaluate ()
+		{
+			bool result = false;
+
+			foreach (var cond in conditions) {
+				bool r = cond.Evaluate ();
+				if (allMustPass && !r) return false;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: Replace the two conditions with this:

```
if (allMustPass ^ r)
    return allMustPass
````

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		public bool IsEnabled => Condition?.Evaluate () ?? true;
+	}
+
+	public static class FeatureSwitchService
+	{
+		static Dictionary<string, FeatureSwitch> features = new Dictionary<string, FeatureSwitch> ();
+
+		public static FeatureSwitch RegisterFeature (string name, FeatureSwitchCondition condition)
+		{
+			var result = features [name] = new FeatureSwitch { Name = name, Condition = condition };
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: Should this be thread safe?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+			return result;
+		}
+
+		public static bool IsEnabled (string name)
+		{
+			if (features.TryGetValue (name, out FeatureSwitch fs)) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: Should this be thread safe?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs (hunk 4 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: I'd probably use an immutabledictionary and retry loops.

```
https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki?wikiVersion=GBwikiMaster&pagePath=%2FTeam%20Pages%2FVS%20for%20Mac%2FCoding%20Guides%2FParallelism%2C%20how%20to%20avoid%20locks.%20Atomic%20operations%20and%20lock%252Dfree%20algorithms
```

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		public EnvVarExistsFeatureSwitchCondition (string name, string value)
+		{
+			nameToCheck = name;
+			valueToCheck = value;
+		}
+
+		internal override bool Evaluate ()
+		{
+			if (!string.IsNullOrEmpty (nameToCheck)) {
+				return string.IsNullOrEmpty (valueToCheck) ? Environment.GetEnvironmentVariables ().Contains (nameToCheck) : Environment.GetEnvironmentVariable (nameToCheck) == valueToCheck;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: Shouldn't this just do GetEnvironmentVariable (nameToCheck) and check `valueToCheck == null || value == valueToCheck`?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+		public static bool IsEnabled (string name)
+		{
+			if (features.TryGetValue (name, out FeatureSwitch fs)) {
+				return fs.IsEnabled;
+			}
+
+			return true; // FIXME: or should we assume feature is disabled if we don't know about it?
+		}
+
+		public static bool IsEnabled (FeatureSwitch fs) => fs?.IsEnabled ?? true;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: Why do we need this?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs (hunk 1 above)
Comment (rodrmoya) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: yes, adding tests in the next commit

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs (hunk 2 above)
Comment (rodrmoya) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs]: yes, indeed, it never changes, as I don't expect the need to add conditions to `Aggregated...` once instantiated, as the idea is to use this like `InstrumentationCounter`'s API, with static fields like in https://github.com/mono/monodevelop/pull/6483/files#diff-563f5ecf5ef6e0a860fd9d77cb55312fR46

**Review by Therzok — APPROVED** [review]

**Review by netonjm — APPROVED** [review]

**Pull request merged**